[2 - 4단계 방탈출 결제 / 배포] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [X] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
객체지향보다 구현에 집중하였습니다.

## 베이스 코드 선택 체크
- [x] 이전 미션의 내 코드에서 시작 
- [ ] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

README에서 설계 시 고민과 작성한 문서를 확인할 수 있습니다!
외부 API 연동 및 테스트, 배포 및 로깅 모두 처음 접하는 영역이라 구현 중 '이렇게 하는 게 맞나' 싶은 생각이 계속 듭니다...
다양한 피드백 남겨주시면 열심히 반영해 보겠습니다! 감사합니다!!

## 배포 서버 URL
- [로그인 페이지](http://ec2-3-35-166-240.ap-northeast-2.compute.amazonaws.com:8080/login)
## API 문서
- [SwaggerHub 링크](https://app.swaggerhub.com/apis/woowacourse-dd6/spring-roomescape_payment_api/1.0.0)
## 데이터베이스 ERD
![spring-roomescape-ERD](https://github.com/user-attachments/assets/d6560470-f94d-41e7-9b2a-51f0ad9047fe))

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jye326 committed:
initial commit : 방탈출 예약 대기 코드 마이그레이션

jye326 committed:
feat(resources) : 토스 결제 위젯 연동 및 예약 로직 수정
- reservation.html: 토스 결제 위젯 스크립트 추가 및 UI 구성 변경
- toss-style.css: 결제창 스타일 정의
- user-reservation.js: 토스 결제 위젯 초기화 및 예약 로직 수정
  - 예약 생성 시 결제 요청 추가
  - fetchReservationPayment 함수로 결제 데이터 처리 로직 구현

jye326 committed:
refactor(js) : 불필요한 주석 제거 및 문서 추가
- user-reservation.js 내 완료된 TODO 주석 제거
- 예약 대기 및 결제 플로우 관련 README.md 추가

jye326 committed:
feat(api): 결제 관련 예약 생성 로직 추가
- ReservationService에 `addAndGetWithoutPayment` 메서드를 사용하도록 변경
- ReservationRequest에 결제 정보를 포함한 필드(paymentKey, orderId, amount, paymentType) 추가

jye326 committed:
feat(service) : 예약 생성 시 결제 승인 로직 추가
- TossPaymentClient를 활용한 결제 승인 로직 추가
- addAndGet 메서드를 addAndGetWithoutPayment와 addAndGet로 분리
- User, ReservationTime, Theme 조회를 위한 메서드 각각 추가 (getUser, getReservationTime, getTheme)
- 관련 테스트 코드 수정

jye326 committed:
feat(payment) : Toss Payments 클라이언트 및 결제 승인 처리 로직 추가
- ClientConfig를 통해 TossPaymentClient 생성 빈 추가
- TossPayments API와 상호작용하는 TossPaymentClient 로직 개발
- 결제 승인 관련 DTO (PaymentApproveDto 및 PaymentApproveResponseDto) 추가

jye326 committed:
feat(exception): 결제 승인 실패에 대한 커스텀 예외 및 에러 코드 추가

jye326 committed:
feat(exception): PaymentApproveException 예외 처리 추가
- PaymentApproveException에 @Getter 적용하여 코드 간소화
- GlobalExceptionHandler에 PaymentApproveException 핸들러 추가

jye326 committed:
refactor(payment): 결제 금액 및 실패 메시지 처리 개선
- 비정상적인 secretKey 값 수정
- 결제 금액(paymentAmount) 테스트 값 조정
- 결제 실패 시 오류 메시지가 alert에 표시되도록 변경

jye326 committed:
feat(config, widget): 결제 금액 및 키 변경
- application.yml에 결제 secret-key 정보 추가
- user-reservation.js에서 결제 금액(paymentAmount) 200,000으로 변경

jye326 committed:
feat(payment): TossPaymentClient refactor 및 secretKey 주입 기능 추가
- ClientConfig에 `@Value`로 secretKey 주입 로직 추가
- TossPaymentClient에 secretKey DI(Dependency Injection) 방식으로 변경
- approvePayment 메서드 내 기존 하드코딩된 secretKey 제거
- RestClient 호출 시 오류 처리 방식 개선 (onStatus 메서드 로직 보완)

jye326 committed:
fix(exception): PaymentApproveException 상속 및 로깅 메시지 수정
- RuntimeException 상속 시 `super` 호출 추가
- 로깅 메시지에서 AuthorizationException을 PaymentApproveException으로 변경

jye326 committed:
test(TossPaymentClient) : Toss Payment 클라이언트 테스트 추가
- 결제 시간이 만료된 경우 예외 처리 테스트
- 잘못된 키 요청 시 예외 처리 테스트

jye326 committed:
refactor(reservation) : 결제 승인 로직 리팩토링 및 불필요한 DTO 제거

- PaymentApproveResponseDto 삭제
- ReservationService 내 `addAndGetWithoutPayment` 메서드 삭제 및 기존 로직 합병
- TossPaymentClient 내 `approvePayment` 메서드 반환값 제거 및 무의미한 DTO 응답 삭제

jye326 committed:
feat(service): 결제 승인 및 예약 생성 테스트 추가
- 결제 승인 요청을 포함한 신규 예약 생성 테스트 케이스 추가
- 대기를 결제 없이 생성하는 테스트 케이스 추가
- TossPaymentClient mock 및 관련 검증 로직 추가

jye326 committed:
feat(payment): TossPaymentClient DI 리팩토링 및 secretKey 관리 방식 개선

- TossPaymentClient 내 secretKey 주입 방식 @Value 애노테이션으로 변경
- ClientConfig 내 TossPaymentClient 생성 로직 단순화 및 RestClient 빈으로 분리
- application.yml에서 결제 secretKey 관리 및 테스트용 설정 추가
- TossPaymentClient의 비효율적인 의존성과 구성 로직 정리

jye326 committed:
test(payment): TossPaymentClient 테스트 개선 및 새로운 케이스 추가

- MockRestServiceServer를 활용한 정상 결제 승인 요청 테스트 추가
- @RestClientTest 애노테이션 사용으로 테스트 환경 설정 개선
- 잘못된 secretKey에 대한 예외 처리 테스트 로직 보완
- 불필요한 하드코딩된 secretKey 제거 및 테스트 데이터 개선
- 연결 시간 및 응답 시간 관련 예외 테스트 주석 추가

jye326 committed:
feat(config): RestClient 연결 및 응답 타임아웃 설정 추가

- ClientConfig에 SimpleClientHttpRequestFactory 설정 추가
- RestClient 요청에 connectTimeout 및 readTimeout 값 적용 (10초로 설정)
- TossPaymentClient 테스트 코드에서 불필요한 주석 제거

jye326 committed:
feat(reservation): 예약 생성 로직에 DTO 추출 및 결제 관리 리팩토링

- `ReservationSpecDto` 및 `PaymentApproveDto` 추가로 메서드 파라미터 단순화
- `addAndGet` 및 `addAndGetWithoutPayment` 로직 통합
- `ReservationSpecDto` 사용으로 예약 생성 관련 데이터 캡슐화
- 결제 승인 관련 로직 개선 및 내구성 강화
- 테스트 코드에서 DTO 변경에 따른 수정 및 리팩토링 적용

jye326 committed:
refactor(config): ClientConfig 패키지 구조 변경 및 테스트 수정

- `ClientConfig`를 `roomescape.infrastructure.payment.config` 패키지로 이동
- `TossPaymentClientTest`에서 새로운 패키지 구조 반영
- 테스트 환경 설정(`@RestClientTest`)에서 `ClientConfig` 추가

jye326 committed:
feat(payment): RestClient 설정 분리 및 TossPaymentClient 리팩토링

- ClientConfig에서 RestClient 생성 로직과 RestClientCustomizer 분리
- TossPaymentClient의 RestClient DI 방식을 Builder로 유연하게 변경
- RestClient 연결 타임아웃(connectTimeout)과 응답 타임아웃(readTimeout) 축소 (10초 → 5초)
- TossPaymentClient 테스트에서 MockRestServiceServer 설정 방식 개선
- TossPaymentClient 생성자에서 불필요한 @Component 애노테이션 제거

jye326 committed:
feat(config): RestClient 연결 및 응답 타임아웃 수정

- ClientConfig에서 connectTimeout 값을 5초 → 3초로 감소
- readTimeout 값을 5초 → 45초로 증가하여 긴 응답 대기 처리 가능하도록 수정

jye326 committed:
feat(payment): 결제 처리 로직 리팩토링 및 PaymentService 도입

- TossPaymentClient를 활용한 결제 승인 로직을 PaymentService로 분리
- Payment 엔티티 및 Repository 추가로 결제 데이터 관리 기능 도입
- ReservationService에서 PaymentService를 사용하여 결제 처리 간소화
- JpaPaymentRepository 및 JpaPaymentDao 추가로 결제 데이터 접근 계층 구성
- TossPaymentClient의 approvePayment 메서드 반환값 변경 및 내구성 강화
- Reservation 생성 로직 내 대기 관련 처리 코드 제거
- 관련 테스트 케이스 수정 및 PaymentService 테스트 추가

jye326 committed:
feat(payment): PaymentApproveResponseDto 추가 및 관련 로직 리팩토링

- TossPaymentClient의 approvePayment 메서드 반환 타입을 PaymentApproveResponseDto로 변경
- Payment 엔티티 생성 로직에서 PaymentApproveResponseDto 사용
- ReservationService 및 PaymentService 내 예약 생성과 결제 처리 간소화
- 불필요한 DTO(PaymentApproveDto) 의존 제거를 통한 코드 정리
- 관련 테스트 코드 수정 및 내구성 강화

jye326 committed:
feat(payment): PaymentApproveRequestDto 도입 및 관련 로직 리팩토링

- 기존 `PaymentApproveDto`를 `PaymentApproveRequestDto`로 변경하여 네이밍 명확화
- ReservationService 및 PaymentService에서 DTO 이름 변경 반영
- TossPaymentClient, PaymentService 등 관련 메서드에 RequestDto 적용
- 중복 예약 검사 기능 `validDuplicatedReservation` 메서드로 추출하여 가독성 개선
- LocalDataInitializer에 결제 데이터 초기화 로직 추가 및 예약-결제 매핑 구현
- 관련 테스트 코드 수정 및 일관성 유지

jye326 committed:
feat(reservation): UserReservationDetailDto 도입 및 예약 조회 리팩토링

- 기존 `ReservationWithAheadDto`를 `UserReservationDetailDto`로 대체하여 네이밍 명확화
- 예약 데이터 조회 시 결제 정보(paymentKey, totalAmount) 포함하도록 수정
- `ReservationRepository` 및 관련 메서드 이름 수정
  - `findReservationsWithAhead` → `findAllReservationDetailByUserId`
- `ReservationApiController.getMyReservations` 메서드에 새로운 DTO 반영
- 관련된 DTO 및 응답 클래스 수정 (`ReservationMineResponse` 등)
- ReservationService, JpaReservationDao 등 비즈니스 및 데이터 계층 반영
- `PaymentApproveRequestDto`, `PaymentApproveResponseDto` 패키지 이동
- 테스트 코드 수정 및 기능 검증

jye326 committed:
feat(reservation): 예약 목록에 결제 정보 표시 기능 추가

- 예약 목록에 paymentKey와 결제 금액(amount) 컬럼 추가
- 예약 데이터 렌더링 로직 변경
  - 결제 정보가 없을 경우 'NOT PAID'로 표시
- 관련 HTML, JavaScript 업데이트

jye326 committed:
merge to step1

jye326 committed:
feat(aop): 컨트롤러 로깅 기능 추가 및 예외 처리 확장

- LogAspect를 통해 AOP 기반 컨트롤러 호출 로깅 기능 구현
  - 메서드 호출 전/후 및 예외 발생 시 로깅 처리
  - API 호출 시 파라미터 및 반환 값 JSON 형태로 로깅
- build.gradle에 spring-boot-starter-aop 의존성 추가
- GlobalExceptionHandler에 MethodArgumentNotValidException 예외 처리 로직 추가
  - 유효성 검사 실패 시 BAD_REQUEST 응답 반환
- 코드 리팩토링 및 관련 로깅 메시지 추가

jye326 committed:
feat(docs): 결제 기능 설계 및 플로우 문서 업데이트

- Step2 결제 도입 설계 문서 추가
  - 요구사항 및 DB 설계 지침 포함
  - 예약과 결제의 분리 및 관계 정의
- 예약 호출 방식 개선 설계
  - LEFT JOIN 활용 조회 방식 제안
  - 반환 DTO 구조 설계 `ReservationPaymentDto`
- 로깅 및 API 문서화 방안 정리
  - 주요 로깅 위치 설정
  - OpenAPI 기반 Swagger 문서 배포 링크 포함
- ERD 다이어그램 추가 (`img.png`)

EastHShin commented on README.md:
```diff
@@ -1 +1,82 @@
+    - openAPI 명세 yaml 파일 생성 : **IntelliJ EndPoint 자동생성 활용**
+    - yaml → SwaggerUI 문서 작성 및 배포 : **swagger Hub 이용**
+
+### ERD
+
+![img.png](img.png)
```
작성해주신 다이어그램은 ERD 가 아닌, 클래스 다이어그램이네요.
ERD에 대해 다시 한번 학습해보시고, 작성해주시면 좋을 것 같아요 :)

jye326 replied:
인텔리제이가 제공하는 ERD 기능을 사용해 보았는데 지금 보니 말씀하신대로 클래스 다이어그램에 가까운 것 같습니다! 다시 작성하여 업로드 하였습니다!
![image (8)](https://github.com/user-attachments/assets/13661cd7-43fb-4373-a102-3ad10b7e7858)
![Untitled (3)](https://github.com/user-attachments/assets/39f96842-fef2-4fb0-9924-ef4137cb0117)

EastHShin replied:
payment, reservation 은 1:1 관계인데, N:1 관계로 표현되고 있네요~
학습하는 단계이니 자동으로 만들어주는 것에 의존하기 보다는 직접 작성해보시는 것도 추천드립니다 :)

EastHShin commented on src/main/java/roomescape/business/model/entity/Payment.java:
```diff
@@ -0,0 +1,53 @@
+    @EmbeddedId
+    private final Id id;
+
+    @OneToOne(optional = false)
+    private Reservation reservation;
```
Payment가 Reservation을 갖도록 구현해주셨네요~
이 반대의 설계도 가능했을 것 같은데, 이렇게 결정해주신 이유는 무엇이었나요?

jye326 replied:
핵심 이유는 결제 유무가 Reservation의 책임이 아니라고 생각했기 때문입니다! 또한 현재 예약 대기 중인 예약을 비롯하여 아직 결제되지 않은 예약도 존재할 수 있고, 이럴 경우 payment는 null값을 갖게 되며 해당 필드에 대한 null 체크가 필요합니다. 따라서 Payment가 결제가 이뤄진 상품을 관리하는 형식으로 구현했습니다!

EastHShin commented on README.md:
```diff
@@ -1 +1,82 @@
+    - [ ] 애플리케이션 시작 및 종료.
+    - [ ] DB작업
+
+### API 문서
+
+- [배포 링크](https://app.swaggerhub.com/apis/woowacourse-dd6/spring-roomescape_payment_api/1.0.0)
+    - openAPI 명세 yaml 파일 생성 : **IntelliJ EndPoint 자동생성 활용**
```
swaggerhub 를 이용해주신 이유가 있나요? 서버 내에 swagger 정적 페이지를 이용할 수 있도록 설정하면 swaggerhub에 따로 배포하지 않아도 되어 편할 것 같네요.

jye326 replied:
Swagger Config 클래스 작성이나 Swagger를 위한 어노테이션을 api 메서드에 붙이는 것이 불필요한 코드로 느껴졌습니다. 인텔리제이가 지원하는 Open API 명세 추출 기능으로 원하는 api손쉽게 yml 파일을 얻을 수 있었고, 이를 클라이언트에 swagger ui 형태로 제공할 수 있는 방법으로 swaggerhub를 선택했습니다!

이후 한 동안 Open API spec 파일로 작성한 커스텀 swagger ui를 서버 내 정적 페이지로 띄우는 작업도 시도해보았으나 과정이 너무 복잡하여 원하는 결과물을 얻는데 실패했습니다.  😥

EastHShin replied:
코드상에 swagger 관련 코드가 침투하는 것이 불필요하게 느껴지셨군요~
물론 그러한 점은 단점이 될 수 있지만, 수동으로 api 문서를 생성하고 swaggerhub 에 업로드하는 것이 오히려 누락할 수 있고 관리가 안될 것 같다고 느껴지네요.

jye326 replied:
넵! 저도 다음에는 Swagger 관련 코드를 추가하는 정석적인(?) 방법을 시도해보려고 합니다!

EastHShin commented on README.md:
```diff
@@ -1 +1,82 @@
+
+- [배포 링크](https://app.swaggerhub.com/apis/woowacourse-dd6/spring-roomescape_payment_api/1.0.0)
+    - openAPI 명세 yaml 파일 생성 : **IntelliJ EndPoint 자동생성 활용**
+    - yaml → SwaggerUI 문서 작성 및 배포 : **swagger Hub 이용**
```
swagger가 아닌 restdocs 를 고려해보시진 않았는지도 궁금하네요~
어떤 기준을 가지고 선택해주셨나요?

jye326 replied:
Open API spec 파일으로 swagger와 restdocs가 제공하는 문서를 각각 확인해 보았습니다. swagger UI가 더 깔끔하고 try 버튼으로 시험 요청 기능을 제공하는 것에 편의를 느껴 선택했습니다~

EastHShin commented on src/main/java/roomescape/exception/GlobalExceptionHandler.java:
```diff
@@ -38,6 +39,13 @@ public ResponseEntity<ErrorResponse> handle(PaymentApproveException e) {
 
+    @ExceptionHandler(MethodArgumentNotValidException.class)
+    public ResponseEntity<ErrorResponse> handleValidationExceptions(MethodArgumentNotValidException e) {
+        logger.warn("Handled MethodArgumentNotValidException: {}", e.getMessage());
```
warn, error, debug, info 레벨을 사용하는 기준이 무엇인가요?

jye326 replied:
warn은 예상할 수 있는 예외, error는 예상할 수 없는 심각한 예외, info는 로직 호출 등의 정보 제공 용으로 사용했습니다. debug는 의미 그대로 디버깅용으로 사용하는 것으로 예상하고 있는데 info와 크게 차이가 없는 것 같아 사용하지 않았습니다!

EastHShin commented on src/main/java/roomescape/log/LogAspect.java:
```diff
@@ -0,0 +1,66 @@
+
+        Object[] args = joinPoint.getArgs();
+        if (args.length <= 0) {
+            log.info("no parameter");
```
이 로그만 덩그러니 남는다면, 여러 요청이 쏟아져 나왔을 때, 어떤 요청에 대해 파라미터가 없었는지 파악하기 어렵겠네요

jye326 replied:
`log.info("api call : {}", method.getName());` 를 통해 요청을 확인할 수 있고, 파라미터가 없는 경우는 어떤 파라미터가 부족한지 파악할 수 있도록 수정했습니다!

EastHShin reviewed (changes requested):
안녕하세요 띠용~

이번 단계도 잘 해주셨네요.
몇가지 코멘트 남겨두었으니 확인 부탁드려요 :)

jye326 committed:
feat(config): Swagger 설정 추가 및 H2 콘솔 JPA 설정 업데이트

- SwaggerConfig.java 추가
  - OpenAPI 기반 API 문서화 설정
  - GroupedOpenApi 및 CustomOpenAPI 설정 포함
- application.yml JPA 설정 변경
  - show-sql 활성화 및 SQL 포맷 포맷팅 설정 추가
- 프로젝트 초기 Swagger UI 설정 파일 추가

jye326 committed:
feat(config): Swagger 설정 제거 및 의존성 버전 업데이트

- SwaggerConfig.java 삭제
  - OpenAPI 및 GroupedOpenApi 설정 제거
  - Swagger 기반 API 문서화 미사용으로 관련 파일 제거
- build.gradle 의존성 버전 업데이트
  - springdoc-openapi-starter-webmvc-ui 버전 2.0.2 → 2.8.8으로 수정
  - 최신 버전 반영 및 의존성 정리

jye326 committed:
feat(log) : 파라미터 부족 에러 로깅

- 부족 파라미터 표시 추가

EastHShin reviewed (approved):
안녕하세요 띠용

미션 진행하시느라 고생하셨습니다~
이번 미션은 여기서 마무리하도록 할게요. 추가 코멘트 확인해주시고 궁금한 점 있으시면 DM 주세요 :)
